examples/initializingworkspaces/main.go

```diff
@@ -0,0 +1,164 @@
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func init() {
+	runtime.Must(corev1alpha1.AddToScheme(scheme.Scheme))
+	runtime.Must(tenancyv1alpha1.AddToScheme(scheme.Scheme))
+	runtime.Must(apisv1alpha1.AddToScheme(scheme.Scheme))
+}
+
+func main() {
+	log.SetLogger(zap.New(zap.UseDevMode(true)))
+	ctx := signals.SetupSignalHandler()
+	entryLog := log.Log.WithName("entrypoint")
+
+	var (
+		server          string
+		initializerName string
+		provider        *initializingworkspaces.Provider
+	)
+
+	pflag.StringVar(&server, "server", "", "Override for kubeconfig server URL")
+	pflag.StringVar(&initializerName, "initializer", "initializer:example", "Name of the initializer to use")
+	pflag.Parse()
+
+	cfg := ctrl.GetConfigOrDie()
+	cfg = rest.CopyConfig(cfg)
+
+	if server != "" {
+		cfg.Host = server
+	}
+	// Setup a Manager, note that this not yet engages clusters, only makes them available.
+	entryLog.Info("Setting up manager")
+	opts := manager.Options{}
+
+	var err error
+	provider, err = initializingworkspaces.New(cfg, initializingworkspaces.Options{InitializerName: initializerName})
+	if err != nil {
+		entryLog.Error(err, "unable to construct cluster provider")
+		os.Exit(1)
+	}
+
+	mgr, err := mcmanager.New(cfg, provider, opts)
+	if err != nil {
+		entryLog.Error(err, "unable to set up overall controller manager")
+		os.Exit(1)
+	}
+
+	if err := mcbuilder.ControllerManagedBy(mgr).
+		Named("kcp-initializer-controller").
+		For(&corev1alpha1.LogicalCluster{}).
+		Complete(mcreconcile.Func(
+			func(ctx context.Context, req mcreconcile.Request) (ctrl.Result, error) {
+				log := log.FromContext(ctx).WithValues("cluster", req.ClusterName)
+
+				cl, err := mgr.GetCluster(ctx, req.ClusterName)
+				if err != nil {
+					return reconcile.Result{}, fmt.Errorf("failed to get cluster: %w", err)
+				}
+				client := cl.GetClient()
+
+				lc := &corev1alpha1.LogicalCluster{}
+				if err := client.Get(ctx, req.NamespacedName, lc); err != nil {
+					return reconcile.Result{}, err
+				}
+
+				// check if your initializer is still set on the logicalcluster
+				if slices.Contains(lc.Status.Initializers, corev1alpha1.LogicalClusterInitializer(initializerName)) {
+					log.Info("Starting to initialize cluster")
+					s := &corev1.ConfigMap{
+						ObjectMeta: ctrl.ObjectMeta{
+							Name:      "kcp-initializer-cm",
+							Namespace: "default",
+						},
+						Data: map[string]string{
+							"test-data": "example-value",
+						},
+					}
+					log.Info("Reconciling ConfigMap", "name", s.Name, "uuid", s.UID)
+					if err := client.Create(ctx, s); err != nil {
+						return reconcile.Result{}, fmt.Errorf("failed to create configmap: %w", err)
+					}
```

Comment:
I think the fact that we are using a core resource here that will be in every workspace by default is hiding an issue with the RESTmapper in the current implementation based on wildcard cache / scoped cluster. Can you maybe extend the example to 

1.) use some resource that isn't core. Could be `Workspaces`, so e.g. if the initializer is on a logical cluster, we create a `Workspace` with a certain name and type in it?
2.) first get the object (to make sure it's not there yet, for example), update it and wait for it to be ready before removing the initializer?

At the moment the example doesn't use the cache at all I think, and we should make sure it's functional.

Reply:
Should we completely drop the core resource in this example?

Reply:
Yes, we don't really need it if instead of creating a ConfigMap we create a Workspace.

Reply:
Thanks for the finding, it is true that the current implementation won't work with initializingworkspaces virtual workspace.